lib/web/apiserver.go

```diff
@@ -2254,6 +2280,86 @@ func (h *Handler) siteNodeConnect(
 	return nil, nil
 }
 
+func (h *Handler) generateSession(ctx context.Context, clt auth.ClientI, req *TerminalRequest, clusterName string) (session.Session, error) {
+	var (
+		host string
+		port int
+	)
+	h.log.Infof("Generating new session for %s\n", clusterName)
+	// req.Server will be either an IP, hostname, or server UUID.
+	_, err := uuid.Parse(req.Server)
+	if err != nil {
+		// TODO Investigate using ListResources instead of GetNodes as we can
+		// provide more filters that may reduce the work to do on the client side.
+		servers, err := clt.GetNodes(ctx, apidefaults.Namespace)
+		if err != nil {
+			return session.Session{}, trace.Wrap(err)
+		}
+		host, port, err = resolveServerHostPort(req.Server, servers)
+		if err != nil {
+			return session.Session{}, trace.Wrap(err)
+		}
+	} else {
+		// If the user has passed in a UUID then query for that node directly. We
+		// can connect to `uuid:0` directly. with `host = req.Server` however we will
+		// not have the proper hostname and the UI will display the UUID instead.
+		// This is why we have to query for the node. In a future update we can query
+		// for a uuid:hostname to dramatically improve performance in large environments.
+		server, err := clt.GetNode(ctx, apidefaults.Namespace, req.Server)
+		if err != nil {
+			return session.Session{}, trace.Wrap(err)
+		}
+		host, port, err = resolveServerHostPort(req.Server, []types.Server{server})
+		if err != nil {
+			return session.Session{}, trace.Wrap(err)
+		}
+	}
+
+	return session.Session{
+		Login:          req.Login,
+		ServerID:       req.Server,
+		ClusterName:    clusterName,
+		ServerHostname: host,
+		ServerHostPort: port,
+		ID:             session.NewID(),
+		Created:        time.Now().UTC(),
+		LastActive:     time.Now().UTC(),
+		Namespace:      apidefaults.Namespace,
+	}, nil
+}
+
+func (h *Handler) fetchExistingSession(ctx context.Context, clt auth.ClientI, req *TerminalRequest, siteName string) (session.Session, string, error) {
+	// Fetch the session data from the supplied SID.
```

Comment:
I think you can remove this comment. You aren't actually fetching a session here, just parsing the session ID, and that's obvious from the name of the function.